File: src/typescript-port/id_mm.ts
// ID_MM.TS
// Ported from ID_MM.C - Memory Manager
// Simplified for browser: we just use JS allocation and let GC handle freeing.

export const BUFFERSIZE = 0x1000;

export interface mminfotype {
    nearheap: number;
    farheap: number;
    EMSmem: number;
    XMSmem: number;
    mainmem: number;
}

export const mminfo: mminfotype = {
    nearheap: 0,
    farheap: 0,
    EMSmem: 0,
    XMSmem: 0,
    mainmem: 0,
};

export let bufferseg: Uint8Array = new Uint8Array(BUFFERSIZE);
export let mmerror = false;

export function MM_Startup(): void {
    bufferseg = new Uint8Array(BUFFERSIZE);
    mminfo.mainmem = 4 * 1024 * 1024;  // Report 4MB available
    mminfo.farheap = 4 * 1024 * 1024;
}

export function MM_Shutdown(): void {
    // Nothing to do - GC handles it
}

export function MM_GetPtr(size: number): Uint8Array {
    return new Uint8Array(size);
}

export function MM_FreePtr(_ptr: Uint8Array | null): void {
    // GC handles it
}

export function MM_SetPurge(_ptr: Uint8Array | null, _purge: number): void {
    // No-op in browser
}

export function MM_SetLock(_ptr: Uint8Array | null, _locked: boolean): void {
    // No-op in browser
}

export function MM_SortMem(): void {
    // No-op
}

export function MM_UnusedMemory(): number {
    return 4 * 1024 * 1024;
}

export function MM_TotalFree(): number {
    return 4 * 1024 * 1024;
}

export function MM_BombOnError(_bomb: boolean): void {
    // No-op
}

export function MM_MapEMS(): void {
    // No-op
}

File: src/love2d-port/id_mm.lua
-- ID_MM.lua
-- Memory Manager - ported from ID_MM.C
-- In Lua, memory is managed by the garbage collector.
-- This module provides the same API as a thin wrapper.

local id_mm = {}

id_mm.SAVENEARHEAP = 0x400
id_mm.SAVEFARHEAP  = 0
id_mm.BUFFERSIZE   = 0x1000
id_mm.MAXBLOCKS    = 700

-- mminfotype
id_mm.mminfo = {
    nearheap = 0,
    farheap  = 0,
    EMSmem   = 0,
    XMSmem   = 0,
    mainmem  = 512 * 1024,  -- Report plenty of memory
}

id_mm.bufferseg = {}  -- Generic buffer
id_mm.mmerror   = false

id_mm.beforesort = nil
id_mm.aftersort  = nil

function id_mm.MM_Startup()
    -- In Lua, memory is managed by GC. Just report available memory.
    id_mm.mminfo.mainmem = 512 * 1024
    -- Initialize bufferseg as a generic byte buffer
    id_mm.bufferseg = {}
    for i = 0, id_mm.BUFFERSIZE - 1 do
        id_mm.bufferseg[i] = 0
    end
end

function id_mm.MM_Shutdown()
    -- Nothing to do
end

function id_mm.MM_MapEMS()
    -- No-op (no EMS in modern systems)
end

function id_mm.MM_GetPtr(size)
    -- Allocate a new table/buffer. Returns the buffer.
    local buf = {}
    return buf
end

function id_mm.MM_FreePtr(ptr)
    -- Let GC handle it
    -- In C this takes memptr*, in Lua we just nil out
end

function id_mm.MM_SetPurge(ptr, purge)
    -- No-op (GC handles everything)
end

function id_mm.MM_SetLock(ptr, locked)
    -- No-op
end

function id_mm.MM_SortMem()
    -- No-op (no memory fragmentation in Lua)
    if id_mm.beforesort then id_mm.beforesort() end
    if id_mm.aftersort then id_mm.aftersort() end
end

function id_mm.MM_ShowMemory()
    -- Debug display - stub
end

function id_mm.MM_UnusedMemory()
    return 256 * 1024
end

function id_mm.MM_TotalFree()
    return 256 * 1024
end

function id_mm.MM_BombOnError(bomb)
    -- No-op
end

function id_mm.MML_UseSpace(segstart, seglength)
    -- No-op
end

return id_mm
